core/dblookupext/historyRepository.go

```diff
@@ -229,16 +229,14 @@ func (hr *historyRepository) RegisterToBlockTracker(blockTracker BlockTracker) {
 	blockTracker.RegisterSelfNotarizedHeadersHandler(hr.onNotarizedBlocks)
 }
 
+// This is called both with "shardID" == "metachain" (regular notarization events),
+// but also with "shardID" IN [0, 1, 2, ...] when blocks produced by the metachain (as a source) are notarized.
 func (hr *historyRepository) onNotarizedBlocks(shardID uint32, headers []data.HeaderHandler, headersHashes [][]byte) {
-	log.Trace("onNotarizedBlocks()", "shardID", shardID, "len(headers)", len(headers))
-
-	if shardID != core.MetachainShardId {
```

Comment:
oNotarizedBlocks will be called:
- (twice) due to line 228 when node receives a header from a different shard than its own
     a) with a list of different shard headers
     b) with a list of own headers
- due to line 229 when node receives a header from own shard
    a) with a list of own headers

I think the condition was added before to not care about shard headers, but only about meta headers, no matter if it was called with a list of own headers or cross headers. 

I am trying to think about the situation when we would want this called also for shard headers, but right now I don't really see any, maybe I'm missing something.

@SebastianMarian please confirm

Reply:
Your description is ok, and this condition was to care about only what meta blocks say (if you are a meta node you only care about your blocks when we talk about setting the final block. And if you are in shard you also care about what meta says about your shard headers when you are trying to set the final block)

Reply:
Fixed as discussed, using the new handler (notifier).